[SPARK-27661][SQL] Add SupportsNamespaces API

rdblue opened the pull request:
## What changes were proposed in this pull request?

This adds an interface for catalog plugins that exposes namespace operations:
* `listNamespaces`
* `namespaceExists`
* `loadNamespaceMetadata`
* `createNamespace`
* `alterNamespace`
* `dropNamespace`

## How was this patch tested?

API only. Existing tests for regressions.

rdblue committed:
Add SupportsNamespaces API.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10544/
Test PASSed.

SparkQA commented:
**[Test build #105264 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105264/testReport)** for PR 24560 at commit [`3aa9ebd`](https://github.com/apache/spark/commit/3aa9ebd92569afd918189959ffdf602438278e39).

rdblue commented:
@mccheah, @jzhuge, @cloud-fan, and @marmbrus, it would be great to get feedback on this early prototype of the namespace API.

I think we may want to split this into two APIs depending on how catalogs handle namespaces. Not all catalogs will explicitly create namespaces. For example, you could think of paths as namespaces for tables in S3, but paths don't exist in S3 without objects so there is no "create" operation. Namespaces exist only if there are tables or objects in them. So we could have separate interfaces for the list and exists calls and another for namespaces that have metadata, like Hive.

SparkQA commented:
**[Test build #105264 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105264/testReport)** for PR 24560 at commit [`3aa9ebd`](https://github.com/apache/spark/commit/3aa9ebd92569afd918189959ffdf602438278e39).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  final class SetProperty implements NamespaceChange `
  * `  final class RemoveProperty implements NamespaceChange `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105264/
Test PASSed.

cloud-fan commented:
I feel separating creating and listing namespaces may make the API too complicated. I think it's OK if some implementation throws an exception in `createNamespace`.

SparkQA commented:
**[Test build #105410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105410/testReport)** for PR 24560 at commit [`3aa9ebd`](https://github.com/apache/spark/commit/3aa9ebd92569afd918189959ffdf602438278e39).

SparkQA commented:
**[Test build #105410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105410/testReport)** for PR 24560 at commit [`3aa9ebd`](https://github.com/apache/spark/commit/3aa9ebd92569afd918189959ffdf602438278e39).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  final class SetProperty implements NamespaceChange `
  * `  final class RemoveProperty implements NamespaceChange `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105410/
Test PASSed.

rdblue commented:
@cloud-fan, any more comments? Should I close this PR?

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #108142 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108142/testReport)** for PR 24560 at commit [`3aa9ebd`](https://github.com/apache/spark/commit/3aa9ebd92569afd918189959ffdf602438278e39).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13245/
Test PASSed.

SparkQA commented:
**[Test build #108142 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108142/testReport)** for PR 24560 at commit [`3aa9ebd`](https://github.com/apache/spark/commit/3aa9ebd92569afd918189959ffdf602438278e39).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  final class SetProperty implements NamespaceChange `
  * `  final class RemoveProperty implements NamespaceChange `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108142/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/NamespaceChange.java:
```diff
@@ -0,0 +1,97 @@
+ * <pre>
+ *   import NamespaceChange._
+ *   val catalog = Catalogs.load(name)
+ *   catalog.namespaces.alterNamespace(ident,
```
I think it should be `catalog.alterNamespace`?

rdblue replied:
Fixed.

cloud-fan commented:
generally looks good. Tow comments:
1. Shall we add `String[] defaultNamespace` as we discussed in the community sync?
2. Shall we add a dummy implementation and the unit tests?

rdblue committed:
Fix doc typo, add defaultNamespace method.

brkyvz commented:
Would love a dummy (test) implementation as an example as reference.

One thing we can do if we add `String[] defaultNamespace` is that when someone calls `listNamespaces`, we call the argumented `listNamespaces` with the defaultNamespace

rdblue committed:
Implement SupportsNamespaces in TestTableCatalog, add tests.

SparkQA commented:
**[Test build #108186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108186/testReport)** for PR 24560 at commit [`ab5d21c`](https://github.com/apache/spark/commit/ab5d21c7fa48cf846ed255ff6e20e23113ca71f5).

rdblue commented:
@cloud-fan & @brkyvz, I've added an implementation of the interfaces to the test catalog and added tests.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13287/
Test PASSed.

brkyvz commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/SupportsNamespaces.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  /**
+   * List top-level namespaces from the catalog.
```
Should this list the top-level namespaces or within the `defaultNamespace`? If it's also just going to return `a`, why does it need to return an array of arrays instead of just an array?

rdblue replied:
It should list top-level namespaces. The default namespace is the one used as Spark's current namespace when the catalog is the default and should not change the behavior of the catalog.

It returns an array of arrays so you can pass each result into `listNamespaces(Array[String])`. `Array[String]` is the type that we consistently use for a namespace, so I think it is correct to return it here.

rdblue replied:
Adding to my comment about listing top-level namespaces: if the default namespace changed the catalog's behavior, we would have to document exactly how it should change the behavior and have a way to configure it.

I like keeping this simple instead, so these methods do the same thing every time and Spark doesn't require the plugin to maintain a current namespace as state.

brkyvz commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/SupportsNamespaces.java:
```diff
@@ -0,0 +1,145 @@
+   * @throws NamespaceAlreadyExistsException If the namespace already exists
+   * @throws UnsupportedOperationException If create is not a supported operation
+   */
+  void createNamespaceMetadata(
```
`createNamespace` instead of `createNamespaceMetadata`

rdblue replied:
I recently changed this and I think it fits better with the semantics that we require. Namespaces are required to exist if they contain a table. We don't check whether the namespace exists on table creation and leave that to the implementation. So it makes more sense if this is create namespace metadata, because the namespace may already exist and we don't want to throw NamespaceAlreadyExists if there isn't already metadata. Does that make sense?

rdblue replied:
See this test: https://github.com/apache/spark/pull/24560/files#diff-dbc34e865cad1087fa8c24afe5e5e1aeR761-R773

Because the table exists, the namespace already exists. It would be surprising to fail the create because it was implicitly created earlier, and it would also be surprising to allow a create when it already exists. Adding "metadata" to the method name makes this more obvious because you can create the namespace metadata.

brkyvz replied:
got it, that's fair.

cloud-fan replied:
Do you mean sometimes `createNamespaceMetadata` is the same as adding properties to an existing namespace?

rdblue replied:
@cloud-fan, look at the test case that I linked to. In that case, the namespace exists because it contains a table. Calling `createNamespaceMetadata` is allowed because it doesn't have metadata associated with it, which makes it exist independent of tables. So, yes, this does function like adding properties the first time.

cloud-fan replied:
This behavior is a little counter-intuitive to me. When I run `mkdir -p /dir1/dir2/dir3` to create a bunch of directories, `mkdir /dir1` fails even if `dir1` was created implicitly.

It also complicates the abstraction. Now namespace has 2 different kinds: created explicitly and created implicitly. The one created explicitly will make `createNamespaceMetadata` fail, and the one created implicitly will make `createNamespaceMetadata` success.

rdblue replied:
@cloud-fan, that's why this was renamed to `createNamespaceMetadata`, not `createNamespace`.

cloud-fan replied:
It seems to me that a simple `createNamespace` method is more useful, instead of a `createNamespaceMetadata` method that needs more document to clarify the behavior.

rdblue replied:
@cloud-fan, if we do that, then do you think implementations that extend `SupportsNamespaces` should fail to create a table if its namespace doesn't exist? I think that's a bad idea -- think about file system tables where directories are namespaces and may not exist.

I understand not liking the method name, but we need some way to solve this problem. Curious about what you'd suggest.

cloud-fan replied:
If you run `CREATE db.tbl ...` in JDBC, it will complain about `db` not exist. This is the same in Spark, Hive, Preso and RDBMS like postgres.

I think it's more common to fail the table creation if the namespace doesn't exist. Even for file systems, `mkdir` fails if the parent directory doesn't exist, unless you add the `-p` option.

We can add a variant of `createTable` method in the future, which is similar to `mkdir -p` and can create the missing namespaces automatically. By default I think table creation should fail if the namespace doesn't exist, to simplify the semantic.

rdblue replied:
@cloud-fan, some sources like JDBC and Hive will throw an exception when the namespace doesn't exist. The problem is that we don't want to require this for all sources where it doesn't make sense, like a catalog that stores tables in an object store. Object stores don't necessarily have the concept of a directory that exists independently (s3 FS wrappers have conventions like `_$folder$` objects).

That's why failing to create a table if the namespace doesn't exist is optional, and creating namespaces implicitly is allowed. This isn't something we can (or should) change.

So the question is how do we handle this in this API? I think that it makes sense to allow namespace creation that is independent of tables -- by creating namespace metadata.

mccheah replied:
Instead of `createNamespaceMetadata` can we make this something like `tryCreateNamespace`, which is not required to throw an exception if the namespace already exists? Or, `getOrCreateNamespace`, for a similar effect?

brkyvz replied:
@rdblue If the namespace already exists (implicitly for that matter), wouldn't you use alterNamespace to add the metadata?

rdblue replied:
@brkyvz, that also works (and has a test in this PR).

@mccheah, I would be surprised if `tryCreateNamespace` and `getOrCreateNamespace` updated the namespace metadata, which I think is part of the intent here.

Another option is `ensureNamespaceExists` without metadata, which would create or do nothing if it already exists. Then metadata could be added using `alterNamespace`. I don't think that is a very good option either because there is no clean create with metadata operation.

cloud-fan replied:
I'm not a fan of "implementation is free to decide the behaviors". When end-users run SQL queries in Spark, Spark should behave deterministically and consistently, instead of telling end-users "you will see different behaviors if you connect to different catalogs".

IMO, `CREATE TABLE` should fail if the namespace doesn't exist, as it's intuitive and commonly seen in other databases. If a catalog stores tables in an object store, and is hard to detect namespace existence, then it should not support namespace, or try harder to figure out a workaround.

We can add an option/SQL syntax to allow `CREATE TABLE` to create namespaces automatically, but it should be the end-users who decide which behavior to pick, not the catalog. This is similar to `mkdir`. Users must add the `-p` option to create non-existing parent directories. It's pretty bad if we say "mkdir may or may not create parent directories, it depends on which Linux distribution you run it".

brkyvz replied:
@cloud-fan But the whole goal of DSV2 is to let the implementation decide the (internal) behavior, but have Spark to enforce the semantics. If you truncate a table with Spark, all Spark needs is that the next time that table is queried that there is no data in the table. It doesn't care if the table (datasource) decided to physically delete all the data or just did a logical deletion.

 - `CREATE TABLE` should fail if the namespace doesn't exist => This makes sense to me. I'd say this is the behavior that users expect. I don't think this would be a blocker if people start to implement object store backed catalogs. Then they can run `CREATE NAMESPACE a.b.c`, and then try the CREATE TABLE again
 - If a namespace already exists, then IMHO CREATE NAMESPACE should throw an error and they should run ALTER NAMESPACE to add the metadata
 - Regarding implicit namespaces in object stores, I feel that this still means that someone created the namespace through some other means (imagine using a GUI to create a MySQL table instead of the shell), and alter namespace should be used to add the metadata.

wdyt @rdblue @mccheah

cloud-fan replied:
@brkyvz when I say "behavior", I don't mean the internal behavior that is invisible to end-users, I actually mean the semantic that is visible to end-users.

brkyvz reviewed (commented):
Thanks for the implementation. Have one question and one suggestion

brkyvz commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/SupportsNamespaces.java:
```diff
@@ -0,0 +1,145 @@
+   * @throws IllegalStateException If the namespace is not empty
+   * @throws UnsupportedOperationException If drop is not a supported operation
+   */
+  boolean dropNamespace(String[] namespace) throws NoSuchNamespaceException;
```
can we not have a boolean flag here called `cascade` that deletes all the Namespaces and Tables under this namespace if it is not empty?

cloud-fan replied:
+1

rdblue replied:
We can consider adding it later. We should decide whether we want to require implementations to build that, or whether that should be done by Spark.

brkyvz replied:
I see what you mean. Spark can list all tables within all namespaces and delete them if cascade = true, and not have the catalog do it. I can see the pros of that. I think something like that opens you up to race conditions too though. This is fine for now, just wanted to bring it up.

rdblue replied:
I'm generally in favor of Spark taking care of operations like these. Otherwise it puts a lot of responsibility on catalog plugins and we want to avoid making those too complex. Required complexity in a plugin will lead to buggy implementations and unreliable behavior across plugins.

SparkQA commented:
**[Test build #108186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108186/testReport)** for PR 24560 at commit [`ab5d21c`](https://github.com/apache/spark/commit/ab5d21c7fa48cf846ed255ff6e20e23113ca71f5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108186/
Test PASSed.

rdblue committed:
Restore createNamespace call and behavior.

rdblue commented:
@cloud-fan, @brkyvz, @mccheah, from the discussion, I think that the rename from `createNamespace` to `createNamespaceMetadata` was a bad idea. I think there is consensus around using `createNamespace` and failing calls if the namespace already exists because a table exists.

I've reverted the rename to `createNamespaceMetadata` and updated the tests. Now, the method is `createNamespace` and the tests validate that the namespace creation fails when a table exists.

Any objections to that behavior?

SparkQA commented:
**[Test build #108595 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108595/testReport)** for PR 24560 at commit [`14de5b6`](https://github.com/apache/spark/commit/14de5b65a7d29f5f63055175a82c593c4a9a5bc1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13699/
Test PASSed.

SparkQA commented:
**[Test build #108595 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108595/testReport)** for PR 24560 at commit [`14de5b6`](https://github.com/apache/spark/commit/14de5b65a7d29f5f63055175a82c593c4a9a5bc1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108595/
Test FAILed.

SparkQA commented:
**[Test build #108631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108631/testReport)** for PR 24560 at commit [`44afb5c`](https://github.com/apache/spark/commit/44afb5ca97a02f2353669bfb16cb7733c9d53b17).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13718/
Test PASSed.

SparkQA commented:
**[Test build #108631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108631/testReport)** for PR 24560 at commit [`44afb5c`](https://github.com/apache/spark/commit/44afb5ca97a02f2353669bfb16cb7733c9d53b17).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108631/
Test PASSed.

brkyvz reviewed (approved):
LGTM. Thanks @rdblue, merging to master!

cloud-fan commented:
A late LGTM

rdblue commented:
Thanks for reviewing, everyone!
